File: app/src/main/java/com/readrops/app/utils/HtmlParser.java
package com.readrops.app.utils;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.readrops.api.localfeed.LocalRSSHelper;
import com.readrops.api.utils.ApiUtils;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.koin.java.KoinJavaComponent;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public final class HtmlParser {

    private static final String TAG = HtmlParser.class.getSimpleName();

    /**
     * Parse the html page to get all rss urls
     *
     * @param url url to request
     * @return a list of rss urls with their title
     */
    public static List<ParsingResult> getFeedLink(String url) {
        List<ParsingResult> results = new ArrayList<>();

        String head = getHTMLHeadFromUrl(url);
        if (head != null) {
            Document document = Jsoup.parse(head, url);

            Elements elements = document.select("link");

            for (Element element : elements) {
                String type = element.attributes().get("type");

                if (LocalRSSHelper.isRSSType(type)) {
                    String feedUrl = element.absUrl("href");
                    String label = element.attributes().get("title");

                    results.add(new ParsingResult(feedUrl, label));
                }
            }

            return results;
        } else {
            return Collections.emptyList();
        }
    }

    @Nullable
    public static String getFaviconLink(@NonNull String url) {
        String favUrl = null;

        String head = getHTMLHeadFromUrl(url);
        if (head == null)
            return null;

        Document document = Jsoup.parse(head, url);
        Elements elements = document.select("link");

        for (Element element : elements) {
            if (element.attributes().get("rel").toLowerCase().contains("icon")) {
                favUrl = element.absUrl("href");
                break;
            }
        }

        return favUrl;
    }

    @Nullable
    private static String getHTMLHeadFromUrl(@NonNull String url) {
        long start = System.currentTimeMillis();

        try {
            Response response = KoinJavaComponent.get(OkHttpClient.class)
                    .newCall(new Request.Builder().url(url).build()).execute();

            if (response.header("Content-Type").contains(ApiUtils.HTML_CONTENT_TYPE)) {
                String body = response.body().string();
                String head = body.substring(body.indexOf("<head"), body.indexOf("</head>"));

                long end = System.currentTimeMillis();
                Log.d(TAG, "parsing time : " + (end - start));

                return head;
            } else {
                return null;
            }
        } catch (Exception e) {
            Log.d(TAG, e.getMessage());
            return null;
        }

    }
}

File: api/src/main/java/com/readrops/api/utils/HtmlParser.kt
package com.readrops.api.utils

import android.nfc.FormatException
import com.readrops.api.localfeed.LocalRSSHelper
import com.readrops.api.utils.ApiUtils.isHtml
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element

data class ParsingResult(
    val url: String,
    val label: String?,
)

object HtmlParser {

    @Throws(FormatException::class)
    suspend fun getFeedLink(url: String, client: OkHttpClient): List<ParsingResult> {
        val document = getHTMLHeadFromUrl(url, client)

        return document.select("link")
            .filter { element ->
                val type = element.attributes()["type"]
                LocalRSSHelper.isRSSType(type)
            }.map {
                ParsingResult(
                    url = it.absUrl("href"),
                    label = it.attributes()["title"]
                )
            }
    }

    fun getFaviconLink(document: Document): String? {
        val links = document.select("link")
            .filter { element -> element.attributes()["rel"].contains("icon") }
            .sortedWith(compareByDescending<Element> {
                it.attributes()["rel"] == "apple-touch-icon"
            }.thenByDescending { element ->
                val sizes = element.attr("sizes")

                if (sizes.isNotEmpty()) {
                    try {
                        sizes.filter { it.isDigit() }
                            .toInt()
                    } catch (e: Exception) {
                        0
                    }
                } else {
                    0
                }
            })

        return links.firstOrNull()
            ?.absUrl("href")
    }

    fun getFeedImage(document: Document): String? {
        return document.select("meta")
            .firstOrNull { element ->
                val property = element.attr("property")
                listOf("og:image", "twitter:image").any { it == property }
            }
            ?.absUrl("content")
    }

    fun getFeedDescription(document: Document): String? {
        return document.select("meta")
            .firstOrNull { element ->
                val property = element.attr("property")
                listOf("og:title", "twitter:title").any { it == property }
            }
            ?.attr("content")
    }

    suspend fun getHTMLHeadFromUrl(url: String, client: OkHttpClient): Document =
        withContext(Dispatchers.IO) {
            client.newCall(
                Request.Builder()
                    .url(url)
                    .build()
            ).execute()
                .use { response ->
                    val body = response.body
                    if (body?.contentType()?.isHtml == true) {
                        val stringBuilder = StringBuilder()
                        var collectionStarted = false

                        while (!body.exhausted()) {
                            val currentLine = body.readUtf8LineStrict()

                            when {
                                currentLine.contains("<head>") -> {
                                    stringBuilder.append(currentLine)
                                    collectionStarted = true
                                }

                                currentLine.contains("</head>") -> {
                                    stringBuilder.append(currentLine)
                                    break
                                }

                                collectionStarted -> {
                                    stringBuilder.append(currentLine)
                                }
                            }
                        }

                        if (!stringBuilder.contains("<head>") || !stringBuilder.contains("</head>")) {
                            throw FormatException("Failed to get HTML head from $url")
                        }

                        Jsoup.parse(stringBuilder.toString(), url)
                    } else {
                        throw FormatException("Response from $url is not a html file")
                    }
                }
        }
}